File: _tools/bump-asset-hashes.py
"""
Cache-bust CSS/JS URLs in live HTML by content hash.
Bump human-readable site build id (CalVer YYYY.MM.DD.N) in HTML markers.

Usage: python _tools/bump-asset-hashes.py [--all]
  --all  also update Archive/**/*.html

Lives under _tools/ so GitHub Pages (Jekyll) will not publish it.
"""
from __future__ import annotations

import hashlib
import re
import sys
from datetime import date
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
VERSION_FILE = Path(__file__).resolve().parent / "site-version.txt"
INCLUDE_ARCHIVE = "--all" in sys.argv

ASSETS = [
    ("style.css", "href"),
    ("colorbox.css", "href"),
    ("sidebar-loader.js", "src"),
]

BUILD_COMMENT = re.compile(r"<!--\s*iancw-build:\s*[^>]*-->")
BUILD_META = re.compile(r'<meta\s+name="site-version"\s+content="[^"]*"\s*/?>')
BODY_BUILD = re.compile(r'(<body\b[^>]*)\s+data-build="[^"]*"')
BODY_OPEN = re.compile(r"(<body\b[^>]*)(>)")


def short_hash(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()[:8]


def list_html_files() -> list[Path]:
    files = sorted(ROOT.glob("*.html"))
    if INCLUDE_ARCHIVE:
        files.extend(sorted((ROOT / "Archive").rglob("*.html")))
    return files


def next_site_version(current: str) -> str:
    today = date.today().strftime("%Y.%m.%d")
    current = current.strip()
    if current.startswith(f"{today}."):
        seq = int(current.rsplit(".", 1)[1])
        return f"{today}.{seq + 1}"
    return f"{today}.1"


def read_site_version() -> str:
    if VERSION_FILE.is_file():
        return VERSION_FILE.read_text(encoding="utf-8").strip()
    return next_site_version("")


def bump_site_version() -> str:
    version = next_site_version(read_site_version())
    VERSION_FILE.write_text(f"{version}\n", encoding="utf-8", newline="\n")
    return version


def apply_site_version(text: str, version: str) -> str:
    comment = f"<!-- iancw-build: {version} -->"
    meta = f'<meta name="site-version" content="{version}">'

    if BUILD_COMMENT.search(text):
        text = BUILD_COMMENT.sub(comment, text)
    else:
        text = re.sub(
            r'(\s*<meta\s+charset="utf-8"\s*/?>)',
            rf"\n\t\t{comment}\n\t\t{meta}\n\t\t<meta charset=\"utf-8\">",
            text,
            count=1,
            flags=re.IGNORECASE,
        )

    if BUILD_META.search(text):
        text = BUILD_META.sub(meta, text)

    if BODY_BUILD.search(text):
        text = BODY_BUILD.sub(rf'\1 data-build="{version}"', text)
    else:
        text = BODY_OPEN.sub(rf'\1 data-build="{version}"\2', text, count=1)

    return text


def main() -> None:
    site_version = bump_site_version()
    print(f"site-version -> {site_version}")

    hashes: dict[str, str] = {}
    for name, _attr in ASSETS:
        fp = ROOT / name
        if not fp.is_file():
            print(f"skip missing {name}")
            continue
        hashes[name] = short_hash(fp)
        print(f"{name} -> {hashes[name]}")

    changed = 0
    for html_path in list_html_files():
        text = html_path.read_text(encoding="utf-8")
        before = text
        for name, attr in ASSETS:
            h = hashes.get(name)
            if not h:
                continue
            esc = re.escape(name)
            pattern = re.compile(
                rf'({attr}\s*=\s*"){esc}(?:\?[^"]*)?(")',
                re.IGNORECASE,
            )
            text = pattern.sub(rf"\1{name}?v={h}\2", text)
        text = apply_site_version(text, site_version)
        if text != before:
            html_path.write_text(text, encoding="utf-8", newline="\n")
            changed += 1
            print(f"updated {html_path.relative_to(ROOT)}")

    print(f"Done. {changed} HTML file(s) updated." if changed else "No HTML changes needed.")


if __name__ == "__main__":
    main()

File: _tools/bump-asset-hashes.js
/**
 * Cache-bust CSS/JS URLs in live HTML by content hash.
 * Usage: node _tools/bump-asset-hashes.js [--all]
 *   --all  also update Archive/**/*.html
 *
 * Lives under _tools/ so GitHub Pages (Jekyll) will not publish it.
 */
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

const root = path.resolve(__dirname, '..');
const includeArchive = process.argv.includes('--all');

const assets = [
	{ file: 'style.css', attr: 'href' },
	{ file: 'colorbox.css', attr: 'href' },
	{ file: 'sidebar-loader.js', attr: 'src' },
];

function shortHash(filePath) {
	const buf = fs.readFileSync(filePath);
	return crypto.createHash('sha256').update(buf).digest('hex').slice(0, 8);
}

function listHtmlFiles() {
	const out = [];
	for (const name of fs.readdirSync(root)) {
		if (name.endsWith('.html')) out.push(path.join(root, name));
	}
	if (includeArchive) {
		const walk = (dir) => {
			for (const ent of fs.readdirSync(dir, { withFileTypes: true })) {
				const p = path.join(dir, ent.name);
				if (ent.isDirectory()) walk(p);
				else if (ent.name.endsWith('.html')) out.push(p);
			}
		};
		const arch = path.join(root, 'Archive');
		if (fs.existsSync(arch)) walk(arch);
	}
	return out;
}

const hashes = {};
for (const a of assets) {
	const fp = path.join(root, a.file);
	if (!fs.existsSync(fp)) {
		console.warn('skip missing', a.file);
		continue;
	}
	hashes[a.file] = shortHash(fp);
	console.log(a.file, '->', hashes[a.file]);
}

let filesChanged = 0;
for (const htmlPath of listHtmlFiles()) {
	let text = fs.readFileSync(htmlPath, 'utf8');
	const before = text;
	for (const a of assets) {
		const h = hashes[a.file];
		if (!h) continue;
		const esc = a.file.replace(/\./g, '\\.');
		const re = new RegExp(
			`(${a.attr}\\s*=\\s*")${esc}(?:\\?[^"]*)?(")`,
			'gi'
		);
		text = text.replace(re, `$1${a.file}?v=${h}$2`);
	}
	if (text !== before) {
		fs.writeFileSync(htmlPath, text, 'utf8');
		filesChanged++;
		console.log('updated', path.relative(root, htmlPath));
	}
}

console.log(filesChanged ? `Done. ${filesChanged} HTML file(s) updated.` : 'No HTML changes needed.');
